File: truezip-kernel/src/main/java/de/schlichtherle/truezip/kernel/NeedsLockRetryException.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truezip.kernel;

import de.truezip.kernel.FsModel;
import javax.annotation.CheckForNull;
import javax.annotation.Nullable;
import javax.annotation.concurrent.Immutable;

/**
 * Indicates that all file system locks need to get released before the
 * operation can get retried.
 *
 * @see    FsSyncController
 * @author Christian Schlichtherle
 */
@Immutable
@SuppressWarnings("serial") // serializing an exception for a temporary event is nonsense!
final class NeedsLockRetryException extends FsControlFlowIOException {

    private static final @Nullable NeedsLockRetryException
            SINGLETON = TRACEABLE ? null : new NeedsLockRetryException();

    static NeedsLockRetryException get(FsModel model) {
        return TRACEABLE    ? new NeedsLockRetryException(model, null)
                            : SINGLETON;
    }

    private NeedsLockRetryException() { }

    private NeedsLockRetryException(FsModel model, @CheckForNull Throwable cause) {
        super(model, null, cause);
    }
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/de/schlichtherle/truevfs/kernel/NeedsLockRetryException.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package de.schlichtherle.truevfs.kernel

import javax.annotation.concurrent._

/** Indicates that all file system locks need to get released before the
  * operation can get retried.
  *
  * @see    SyncController
  * @author Christian Schlichtherle
  */
@Immutable
private final class NeedsLockRetryException private ()
extends ControlFlowException

private object NeedsLockRetryException {
  import ControlFlowException._

  private[this] val instance = new NeedsLockRetryException

  def apply() = if (traceable) new NeedsLockRetryException else instance
}
